### ISSUE TITLE
ttnn.group_norm silently wrong for non-tile-aligned H*W when its input's tile padding is non-zero

### ISSUE THREAD

**Issue by ashokkumarkannan1** [tenstorrent/tt-metal#52685]
Follow-up to #51159 (which fixed #50682).

**Two defects are recorded here.** The first is the subject of this issue; the second is a distinct bug in another op, included because it constrains the fix for the first.

1. **`ttnn.group_norm`** is silently wrong for non-tile-aligned `H*W` when its input's tile padding is non-zero — both interleaved and block-sharded, the latter catastrophically. *Fixed by the masking work tracked here.*
2. **`ttnn.fill_implicit_tile_padding`** corrupts BLOCK_SHARDED tensors. **Out of scope for this issue and NOT fixed by masking** — masking merely removes `group_norm`'s need to call it. Any other caller is still affected, so this needs its own fix.

### Observed

The #50682 correction reduces over the tile-padding rows and back-corrects for them **assuming those rows are zero**. Nothing in ttnn guarantees that, so `ttnn.group_norm` is silently wrong for non-tile-aligned `H*W` whenever its input's tile padding is non-zero. No exception is raised.

Non-zero padding is normal, and for some producers it is *correct*:

| producer | padding it leaves |
|---|---|
| `exp` | exactly `1.0` — this is `exp(0)`, the correct elementwise result |
| `softmax` | uniform non-zero weight |
| `slice` | the sliced-away source rows, still physically present |
| `reshape` | source spill / stale bytes (shape-dependent) |

Only host-upload paths (`from_torch`, `tilize_with_zero_padding`) reliably zero it — and **every existing group_norm test builds its input that way**, which is why the whole suite passes against this bug.

A producer-side fix is not viable: `reshape` deliberately only fills padding when the caller passes `pad_value` (a default-on fill was found to clobber logical lanes of zero-cost-aliased buffers — see the fill rules in `reshape.cpp`), and for `exp` there is nothing to fix.

### Error (no exception — silently wrong output)

**Interleaved**, `GroupNorm(num_groups=32, C=1024)`, `H*W=259` (pads to 288). Identical data, only the padding rows differ:

```
padding = 0.0      maxErr=0.0960  PCC=1.00000  worst per-group std ratio=0.972 (want 1.000)
padding = 0.5      maxErr=0.1903  PCC=0.99989  worst per-group std ratio=0.958
padding = 2.0      maxErr=0.7338  PCC=0.99999  worst per-group std ratio=0.834
padding = 7.0      maxErr=2.3203  PCC=0.99997  worst per-group std ratio=0.402
```

**PCC does not detect this** (stays >=0.9998 while per-group std collapses to 0.40) — the contamination is a per-group affine transform, the same blindness as #50682.

**Block-sharded is far worse**, `C=256`, `H*W=100` (pads to 128), grid 2x2. Dirty padding survives the interleaved -> sharded conversion intact:

```
padding = zeros    maxErr=     0.0630   worst per-group std ratio=0.976
padding = 7.0      maxErr=     2.2920   worst per-group std ratio=0.094
padding = 1.0      maxErr=786430.25     worst per-group std ratio=0.000
```

At `1.0` the corrected variance crosses zero and `rsqrt` explodes — a blow-up, not a degradation. **`1.0` is exactly what `exp()` leaves**, so the worst case is reachable from an ordinary `exp -> group_norm` graph.

Real-model impact — XTTS-v2 conditioning encoder (`GroupNorm(32, 1024)`, mel length 259), fed from `reshape`: encoder PCC **0.9996 -> 0.8223**, max err 9.18. Zero-filling that same tensor's padding restores it.

The Welford path is **not** affected: `group_norm` already forces `use_welford=false` for non-tile-aligned `H*W`.

### Expected

`ttnn.group_norm` must be correct for any legal input, independent of its producer's tile-padding contents.

The padding's contribution has to be either made zero or excluded — unknown values cannot be back-corrected. Excluding it (**masking rows `>= logical_hw` in the kernel, in BOTH accumulation passes**) is the only approach that also covers sharded, since it never touches the input tensor and so is layout-independent. It additionally removes the `Var - K*E[x]^2` subtraction that produces the 786430 above.

Masking only the variance pass is **not** sufficient: pass 1 sums the padding rows before centering, so the mean is contaminated too and rescaling the divisor cannot recover it.

The mechanism still needs design work: the correction lives in three program factories (`mcast`, `no_mcast`, `sharded`), and it is worth checking whether other kernels that reduce over a padded dim and back-correct share this assumption.

Whatever lands needs a test that supplies **dirty** padding — every current test zeroes it, so none of them guard this.

#### Defect 2 — `ttnn.fill_implicit_tile_padding` corrupts BLOCK_SHARDED tensors (separate; NOT fixed by masking)

Zero-filling the input is a viable fix for **interleaved** only, and this is why. Feeding a BLOCK_SHARDED tensor (`C=256`, `H*W=100`, grid 2x2) with padding seeded to 7.0 through `fill_implicit_tile_padding(t, 0.0)`:

```
                                      logical data err   padding after fill
BLOCK_SHARDED, before fill                     0.0000           7.000
BLOCK_SHARDED, after  fill(0)                  0.9961           7.000   <-- both contracts broken
INTERLEAVED control, after fill(0)             0.0000           0.000
```

It corrupts the logical region **and** fails to zero the padding, silently, while the interleaved equivalent is correct. This affects **any** caller of that public op, not just `group_norm`.

Masking does not fix this — it only removes `group_norm`'s need to call it. **This defect requires its own fix and must not be considered resolved when the masking work lands.**

### Steps (exact commands)

Save as `repro_group_norm_dirty_padding.py` and run with the tt-metal Python env:

```bash
python repro_group_norm_dirty_padding.py
```

```python
# ttnn.group_norm's non-tile-aligned correction is only correct when the implicit
# tile-padding rows are zero. Identical data, only the padding differs.
import math

import torch
import ttnn

C, HW, G = 1024, 259, 32
PAD = math.ceil(HW / 32) * 32  # 288


def pcc(a, b):
    a, b = a.flatten().float(), b.flatten().float()
    a, b = a - a.mean(), b - b.mean()
    d = a.norm() * b.norm()
    return float((a @ b) / d) if d > 0 else float("nan")


def main():
    dev = ttnn.open_device(device_id=0)
    torch.manual_seed(0)

    x = torch.randn((1, C, 1, HW), dtype=torch.bfloat16)
    ref = (torch.nn.functional.group_norm(x.float(), G)
           .permute(0, 2, 3, 1).reshape(1, 1, HW, C))
    xin = x.permute(0, 2, 3, 1).reshape(1, 1, HW, C)

    print(f"GroupNorm(num_groups={G}, C={C}), H*W={HW} -> tile-padded to {PAD}")
    for fill in (0.0, 0.5, 2.0, 7.0):
        t = ttnn.from_torch(xin, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                            device=dev, memory_config=ttnn.DRAM_MEMORY_CONFIG)
        t = ttnn.fill_implicit_tile_padding(t, fill)
        o = ttnn.group_norm(t, num_groups=G, inplace=False,
                            memory_config=ttnn.DRAM_MEMORY_CONFIG)
        ttnn.synchronize_device(dev)
        out = ttnn.to_torch(ttnn.from_device(o)).float().reshape(1, 1, HW, C)
        ratio = (out.reshape(1, HW, G, C // G).std(dim=(1, 3), unbiased=False) /
                 ref.reshape(1, HW, G, C // G).std(dim=(1, 3), unbiased=False))
        print(f"  padding = {fill:<4} maxErr={(out - ref).abs().max():7.4f}  "
              f"PCC={pcc(ref, out):.5f}  worst per-group std ratio={ratio.min():.3f}")
    ttnn.close_device(dev)


if __name__ == "__main__":
    main()
```

### Software Versions

- Ubuntu 24.04
- tt-metal @ `1a28ab18deefcb9bca72be573e081a31933a7b64` (includes #51159, merged as `ecc15a361295c31b579e2a8408e6425bc3f43948`)
- Reproduced on Wormhole B0 and Blackhole
- Surfaced from tt-xla XTTS-v2 bring-up (tenstorrent/tt-xla#5376) via tt-mlir's group_norm decomposition gate (tenstorrent/tt-mlir#8935). The gate relaxation that would route models onto the fused path is parked on branch `groupnorm-gate-relax` pending this issue.
